plugins/kapt3/kapt3-base/src/org/jetbrains/kotlin/kapt3/base/incremental/cache.kt

```diff
@@ -19,8 +19,18 @@ class JavaClassCacheManager(val file: File) : Closeable {
     private var closed = false
 
     fun updateCache(processors: List<IncrementalProcessor>, failedToAnalyzeSources: Boolean) {
-        if (failedToAnalyzeSources || !aptCache.updateCache(processors)) {
+        if (!aptCache.updateCache(processors, failedToAnalyzeSources)) {
             javaCache.invalidateAll()
+            return
+        }
+        // Compilation is fully incremental, record types defined in generated .class files
+        processors.forEach { processor ->
+            processor.getGeneratedClassFilesToTypes().forEach { (classFile, type) ->
+                val typeInformation = SourceFileStructure(classFile.toURI()).also {
+                    it.addDeclaredType(type)
+                }
+                javaCache.addSourceStructure(typeInformation)
```

Comment:
NB: `JavaClassCache.invalidateTypesForFiles ` is unused now

When class files entries are deleted from cache when AP origins are not presented any more in sources

Reply:
Looks like it's `javaCache.invalidateDataForTypes(impactedTypes)`

Reply:
I'll remove `invalidateTypesForFiles`. Correct `invalidateDataForTypes` is used to remove `.class` (and `.java`) file type information.